### PR TITLE
planner: add more test cases for join order hint

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/29932

Problem Summary:
add more test cases for join order hint and add some warning messages for some forbidden cases,

### What is changed and how it works?
add more test cases for join order hint and add some warning messages for some forbidden cases,

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/rule_join_reorder_test.go [hunk 1]
```diff
@@ -15,6 +15,8 @@
 package core_test
 
 import (
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/parser/model"
 	"testing"
```
Comment (qw4990) [inline planner/core/rule_join_reorder_test.go:20]: Please format these lines.

File: planner/core/rule_join_reorder.go [hunk 2]
```diff
@@ -181,12 +177,30 @@ func (s *joinReOrderSolver) optimizeRecursive(ctx sessionctx.Context, p LogicalP
...
-		if hintInfo != nil && hintInfo.leadingJoinOrder != nil {
+		leadingHintNum := len(hintInfo)
+		var leadingHintInfo *tableHintInfo
+		if leadingHintNum > 0 {
+			hasDiffLeadingHint := false
+			leadingHintInfo = hintInfo[0]
+			// One join group has one leading hint at most. Check whether there are different join order hints.
+			for i := 1; i < leadingHintNum; i++ {
+				if hintInfo[i] != hintInfo[i-1] {
+					hasDiffLeadingHint = true
```
Comment (qw4990) [inline planner/core/rule_join_reorder.go:188]: Why check it here again, isn't it duplicated with `logical_plan_builder.go:L3633`?

Comment (Reminiscent) [reply planner/core/rule_join_reorder.go:188]: `logical_plan_builder.go:L3633` check the leading hints in the same query block.

For here, some leading hints in different query block can appear in the same join group. Such as, `select /*+ leading(t3) */ * from (select /*+ leading(t1) */ t2.b from t1 join t2 on t1.a=t2.a) t4 join t3 on t4.b=t3.b`

**Review by qw4990 — APPROVED** [review]

File: planner/core/testdata/join_reorder_suite_out.json [hunk 3]
```diff
@@ -2074,6 +2074,2503 @@
...
+          "      │ └─Selection 9980.01 cop[tikv]  not(isnull(test.t1.a)), not(isnull(test.t1.b))",
+          "      │   └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo",
+          "      └─TableReader(Probe) 9990.00 root  data:Selection",
+          "        └─Selection 9990.00 cop[tikv]  not(isnull(test.t.a))",
+          "          └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Warning": null
+      },
+      {
+        "SQL": "select /*+ leading(t2) hash_join(t1) */ * from t, t1, t2, t3 where t.a = t1.a and t1.b=t2.b;",
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:2163]: Need further check here

File: planner/core/testdata/join_reorder_suite_out.json [hunk 4]
```diff
@@ -2074,6 +2074,2503 @@
...
+          "      │ └─Selection 9980.01 cop[tikv]  not(isnull(test.t1.a)), not(isnull(test.t1.b))",
+          "      │   └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo",
+          "      └─TableReader(Probe) 9990.00 root  data:Selection",
+          "        └─Selection 9990.00 cop[tikv]  not(isnull(test.t.a))",
+          "          └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Warning": null
+      },
+      {
+        "SQL": "select /*+ leading(t2) INL_JOIN(t1) */ * from t, t1, t2, t3 where t.a = t1.a and t1.b=t2.b;",
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:2184]: ditto

File: planner/core/testdata/join_reorder_suite_out.json [hunk 5]
```diff
@@ -2074,6 +2074,2503 @@
...
+          "          │ └─Selection 9980.01 cop[tikv]  not(isnull(test.t1.a)), not(isnull(test.t1.b))",
+          "          │   └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo",
+          "          └─TableReader(Probe) 9990.00 root  data:Selection",
+          "            └─Selection 9990.00 cop[tikv]  not(isnull(test.t.a))",
+          "              └─TableFullScan 10000.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ],
+        "Warning": null
+      },
+      {
+        "SQL": "select /*+ leading(t2) INL_JOIN(t1) */ * from t, t1, t2, t3 where t.a = t1.a and t1.b=t2.b;",
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:2255]: ditto

File: planner/core/testdata/join_reorder_suite_out.json [hunk 6]
```diff
@@ -2074,6 +2074,2503 @@
...
+          "    │   └─TableFullScan 10000.00 cop[tikv] table:t2, partition:p2 keep order:false, stats:pseudo",
+          "    ├─TableReader 9990.00 root  data:Selection",
+          "    │ └─Selection 9990.00 cop[tikv]  not(isnull(test.t2.b))",
+          "    │   └─TableFullScan 10000.00 cop[tikv] table:t2, partition:p3 keep order:false, stats:pseudo",
+          "    └─TableReader 9990.00 root  data:Selection",
+          "      └─Selection 9990.00 cop[tikv]  not(isnull(test.t2.b))",
+          "        └─TableFullScan 10000.00 cop[tikv] table:t2, partition:p4 keep order:false, stats:pseudo"
+        ],
+        "Warning": [
+          "Warning 1815 leading hint is inapplicable, check if the leading hint table is valid"
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:2444]: ditto

File: planner/core/testdata/join_reorder_suite_out.json [hunk 7]
```diff
@@ -2074,6 +2074,2503 @@
...
+          "  │ └─Selection 9990.00 cop[tikv]  not(isnull(test.t2.a))",
+          "  │   └─TableFullScan 10000.00 cop[tikv] table:t2 keep order:false, stats:pseudo",
+          "  └─TableReader(Probe) 9990.00 root  data:Selection",
+          "    └─Selection 9990.00 cop[tikv]  not(isnull(test.t1.a))",
+          "      └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo"
+        ],
+        "Warning": null
+      },
+      {
+        "SQL": "select /*+ straight_join() */ * from t1 join t2 on t1.a=t2.a right join t3 on t2.b=t3.b",
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:2648]: ditto

Comment (Reminiscent) [reply planner/core/testdata/join_reorder_suite_out.json:2648]: correct?

File: planner/core/testdata/join_reorder_suite_out.json [hunk 8]
```diff
@@ -2074,6 +2074,2503 @@
...
+          "HashJoin 1000000000000.00 root  CARTESIAN inner join",
+          "├─TableReader(Build) 10000.00 root  data:TableFullScan",
+          "│ └─TableFullScan 10000.00 cop[tikv] table:t3 keep order:false, stats:pseudo",
+          "└─HashJoin(Probe) 100000000.00 root  CARTESIAN inner join",
+          "  ├─TableReader(Build) 10000.00 root  data:TableFullScan",
+          "  │ └─TableFullScan 10000.00 cop[tikv] table:t2 keep order:false, stats:pseudo",
+          "  └─TableReader(Probe) 10000.00 root  data:TableFullScan",
+          "    └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo"
+        ],
+        "Warning": null
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:2795]: Lack warning for the straight_join.

File: planner/core/testdata/join_reorder_suite_out.json [hunk 9]
```diff
@@ -2074,6 +2074,2503 @@
...
+          "  │ └─TableFullScan 10000.00 cop[tikv] table:t3 keep order:false, stats:pseudo",
+          "  └─HashJoin(Probe) 12487.50 root  inner join, equal:[eq(test.t1.a, test.t2.a)]",
+          "    ├─TableReader(Build) 9990.00 root  data:Selection",
+          "    │ └─Selection 9990.00 cop[tikv]  not(isnull(test.t2.a))",
+          "    │   └─TableFullScan 10000.00 cop[tikv] table:t2 keep order:false, stats:pseudo",
+          "    └─TableReader(Probe) 9990.00 root  data:Selection",
+          "      └─Selection 9990.00 cop[tikv]  not(isnull(test.t1.a))",
+          "        └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo"
+        ],
+        "Warning": null
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:3062]: Lack warning

File: planner/core/testdata/join_reorder_suite_out.json [hunk 10]
```diff
@@ -2074,6 +2074,2503 @@
...
+          "  │ └─Selection 9990.00 cop[tikv]  not(isnull(test.t2.a))",
+          "  │   └─TableFullScan 10000.00 cop[tikv] table:t2 keep order:false, stats:pseudo",
+          "  └─HashJoin(Probe) 99900000.00 root  CARTESIAN inner join",
+          "    ├─TableReader(Build) 9990.00 root  data:Selection",
+          "    │ └─Selection 9990.00 cop[tikv]  not(isnull(test.t1.a))",
+          "    │   └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo",
+          "    └─TableReader(Probe) 10000.00 root  data:TableFullScan",
+          "      └─TableFullScan 10000.00 cop[tikv] table:t3 keep order:false, stats:pseudo"
+        ],
+        "Warning": null
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:3113]: ditto

File: planner/core/testdata/join_reorder_suite_out.json [hunk 11]
```diff
@@ -2074,6 +2074,2503 @@
...
+          "  │ └─Selection 9990.00 cop[tikv]  not(isnull(test.t2.a))",
+          "  │   └─TableFullScan 10000.00 cop[tikv] table:t2 keep order:false, stats:pseudo",
+          "  └─TableReader(Probe) 9990.00 root  data:Selection",
+          "    └─Selection 9990.00 cop[tikv]  not(isnull(test.t1.a))",
+          "      └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo"
+        ],
+        "Warning": null
+      },
+      {
+        "SQL": "select /*+ straight_join() */ * from t1 join t2 on t1.a=t2.a where exists (select t3.a from t3)",
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:3385]: Need to check plan in the master

File: planner/core/testdata/join_reorder_suite_out.json (hunk 3 above)
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:2163]: It needs to be marked in the document.

File: planner/core/testdata/join_reorder_suite_out.json [hunk 12]
```diff
@@ -2074,6 +2074,2503 @@
...
+          "  └─IndexLookUp(Probe) 1.25 root  ",
+          "    ├─Selection(Build) 1.25 cop[tikv]  not(isnull(test.t1.a))",
+          "    │ └─IndexRangeScan 1.25 cop[tikv] table:t1, index:a(a) range: decided by [eq(test.t1.a, test.t.a)], keep order:false, stats:pseudo",
+          "    └─Selection(Probe) 1.25 cop[tikv]  not(isnull(test.t1.b))",
+          "      └─TableRowIDScan 1.25 cop[tikv] table:t1 keep order:false, stats:pseudo"
+        ],
+        "Warning": null
+      },
+      {
+        "SQL": "select /*+ leading(t1) hash_join(t1) */ * from t, t1, t2, t3 where t.a = t1.a and t1.b=t2.b;",
```
Comment (Reminiscent) [inline planner/core/testdata/join_reorder_suite_out.json:2142]: change table from `t1` to `t2`

File: planner/core/testdata/join_reorder_suite_out.json (hunk 6 above)
Comment (Reminiscent) [reply planner/core/testdata/join_reorder_suite_out.json:2451]: In `static` mode, we will use the table through `ParitionUnion`. The table alias for `PartitionUnion` is nil. So the table name in the leading hint can not be identified.

File: planner/core/testdata/join_reorder_suite_out.json [hunk 13]
```diff
@@ -2841,7 +2851,10 @@
           "  └─TableReader(Probe) 10000.00 root  data:TableFullScan",
           "    └─TableFullScan 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo"
         ],
-        "Warning": null
+        "Warning": [
+          "Warning 1815 leading hint is inapplicable, check the join type or the join algorithm hint",
```
[the remark this answers is not in the document]
Comment (Reminiscent) [reply planner/core/testdata/join_reorder_suite_out.json:2855]: Not easy to optimize here. The number of warning messages depends on the tableName in the leading hint.

**Review by winoros — APPROVED** [review]

**Pull request merged**